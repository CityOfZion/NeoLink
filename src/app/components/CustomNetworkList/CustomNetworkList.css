.networkWrapper {
  display: flex;
  width: 100%;
}

.deleteIcon {
  display: flex;
  align-items: center;
  padding-left: 10px;
  margin-left: auto;
  cursor: pointer;
}

.networkWrapper div.networkInfo {
  display: block;
  word-wrap: break-word;
  font-size: 11px;
}