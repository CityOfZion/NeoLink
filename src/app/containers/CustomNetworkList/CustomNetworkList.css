.manageNetworksContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
}

.manageNetworksHeading {
  margin-top: 20px;
}

.customNetworkDropdown {
  margin: 0;
  padding: 0;
  list-style-type: none;
  width: 80px;
}

.customNetworkDropDownButton,
.customNetworkLink {
  width: 100%;
  height: 50px;
  color: #585858;
  border: none;
  background: transparent;
  font-size: 0.8rem;
  cursor: pointer;
  display: block;
  width: 100%;
  height: 35px;
}

.customNetworkLink {
  display: flex;
  padding: 5px 12px;
  text-decoration: none;
  align-items: center;
}

.customNetworkLink svg {
  margin-right: 3px;
}

.customNetworkDropDownButton:hover,
.customNetworkLink:hover {
  background-color: #ededed;
}

.customNetworkNoNetworksBox {
  width: 80%;
  background-color: #fff;
  color: #585858;
  border-radius: 3px;
  padding: 20px;
  text-align: center;
  font-size: 0.9rem;
  display: flex;
  flex-direction: column;
  align-items: center;
}
