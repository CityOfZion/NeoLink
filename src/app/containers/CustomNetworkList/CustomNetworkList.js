import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import BackNavigation from '../../components/BackNavigation'
import CustomNetworkCard from '../../components/CustomNetworkCard'
import Overlay from '../../components/Overlay'
import ConfirmDelete from '../../components/ConfirmDelete'
import PrimaryButton from '../../components/common/buttons/PrimaryButton'

import trashSVG from '../../../img/trash.svg'
import editSVG from '../../../img/edit.svg'

import style from './CustomNetworkList.css'

// import { truncateUrl } from '../../utils/api/neon'

class CustomNetworkList extends Component {
  state = {
    showConfirmDelete: false,
    currentItem: {},
  }

  delete = index => {
    const { deleteCustomNetwork, setNetwork, selectedNetworkId } = this.props

    // If this is the current network they're using, reset to MainNet
    if (selectedNetworkId === index) {
      setNetwork('MainNet')
    }

    deleteCustomNetwork(index)
  }

  _generateDropDownContent = (index, name) => (
    <ul className={ style.customNetworkDropdown }>
      <li>
        <button
          className={ style.customNetworkDropDownButton }
          onClick={ () => {
            this.showConfirmDelete()
            this.setCurrentItem(index, name)
          } }
        >
          <img src={ trashSVG } alt='trashcan' className={ style.customNetworkDropDownButtonImage } /> Delete
        </button>
        <Link to={ `/editCustomNetwork/${name}` } className={ style.customNetworkLink }>
          <img src={ editSVG } alt='pen on paper' className={ style.customNetworkDropDownButtonImage } />Edit
        </Link>
      </li>
    </ul>
  )

  showConfirmDelete = () => this.setState({ showConfirmDelete: true })

  setCurrentItem = (index, name) => {
    const currentItem = { name, index }
    this.setState({ currentItem })
  }

  handleDelete = index => {
    this.delete(index)
    this.setState({ showConfirmDelete: false })
  }

  generateNetworkRows(networks) {
    const networkRows = []
    Object.keys(networks).forEach(index => {
      const network = networks[index]

      if (network.canDelete) {
        networkRows.push(
          <CustomNetworkCard
            name={ network.name }
            url={ network.url }
            key={ network.name }
            dropDownContent={ this._generateDropDownContent(index, network.name) }
          />
        )
      }
    })

    return networkRows
  }

  render() {
    const { showConfirmDelete, currentItem } = this.state
    const { networks, history } = this.props

    const networkRows = this.generateNetworkRows(networks)

    const content = networkRows.length ? (
      <Fragment>{networkRows}</Fragment>
    ) : (
      <div className={ style.customNetworkNoNetworksBox }>
        <h4>You have no custom networks.</h4>
        <PrimaryButton buttonText={ 'Add Network' } onClickHandler={ () => history.push('/addCustomNetwork') } />
      </div>
    )

    return (
      <div>
        {showConfirmDelete && (
          <Overlay>
            <ConfirmDelete
              onClickAcceptHandler={ () => this.handleDelete(currentItem.index) }
              onClickRejectHandler={ () => this.setState({ showConfirmDelete: false }) }
              item={ currentItem.name }
            />
          </Overlay>
        )}
        <BackNavigation onClickHandler={ () => history.push('/settings') } />
        <section className={ style.manageNetworksContainer }>
          <h1 className={ style.manageNetworksHeading }>Manage Networks</h1>
          {content}
        </section>
      </div>
    )
  }
}

CustomNetworkList.propTypes = {
  networks: PropTypes.object.isRequired,
  deleteCustomNetwork: PropTypes.func.isRequired,
  setNetwork: PropTypes.func.isRequired,
  selectedNetworkId: PropTypes.string.isRequired,
  history: PropTypes.object.isRequired,
}

export default CustomNetworkList
