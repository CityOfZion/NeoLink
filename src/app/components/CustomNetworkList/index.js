import CustomNetworkList from './CustomNetworkList'
export default CustomNetworkList
