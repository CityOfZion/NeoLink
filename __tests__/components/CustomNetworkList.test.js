import React from 'react'

import { shallow } from 'enzyme'

import CustomNetworkList from '../../src/app/components/CustomNetworkList'

const setup = (networkId = 0) => {
  const props = {
    networks: [
      { name: 'MainNet', url: 'http://api.wallet.cityofzion.io', canDelete: false },
      { name: 'local', url: 'http://127.0.0.1:5000', canDelete: true },
    ],
    network: { id: networkId },
    deleteCustomNetwork: jest.fn(),
    setNetwork: jest.fn(),
  }
  const wrapper = shallow(<CustomNetworkList { ...props } />)

  return {
    wrapper,
  }
}

describe('CustomNetworkList', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })

  test('lists networks properly', async () => {
    const { wrapper } = setup()

    // MainNet shouldn't show on this list, as it's not a custom network (canDelete = false).
    expect(wrapper.contains('MainNet')).toEqual(false)

    // Local is a custom network and should show.
    expect(wrapper.contains('local')).toEqual(true)
  })

  test('delete network works', async () => {
    const { wrapper } = setup(1)
    const instance = wrapper.instance()

    // MainNet shouldn't show on this list, as it's not a custom network (canDelete = false).
    expect(wrapper.contains('MainNet')).toEqual(false)

    // Local is a custom network and should show.
    expect(wrapper.contains('local')).toEqual(true)

    wrapper.find('a').simulate('click')
    expect(instance.props.setNetwork).toHaveBeenCalledWith(0)
    expect(instance.props.deleteCustomNetwork).toHaveBeenCalledWith(1)
  })
})
