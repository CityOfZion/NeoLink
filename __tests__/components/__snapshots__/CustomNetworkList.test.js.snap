// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomNetworkList renders without crashing 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CustomNetworkList
      config={
        Object {
          "selectedNetworkId": "MainNet ",
        }
      }
      deleteCustomNetwork={[MockFunction]}
      history={Object {}}
      networks={
        Object {
          "Local": Object {
            "canDelete": true,
            "name": "local",
            "url": "http://127.0.0.1:5000",
          },
          "MainNet": Object {
            "canDelete": false,
            "name": "MainNet",
            "url": "http://api.wallet.cityofzion.io",
          },
        }
      }
      onDeleteClickHandler={[MockFunction]}
      selectedNetworkId="MainNet"
      setNetwork={[MockFunction]}
    >
      <div>
        <SettingsNavigation
          history={Object {}}
        >
          <section
            className="settingsNavigation"
          >
            <button
              className="settingsNavigationButton"
              onClick={[Function]}
            >
              <i
                className="fa fa-arrow-left"
              />
              Back
            </button>
          </section>
        </SettingsNavigation>
        <section
          className="manageNetworksContainer"
        >
          <h1
            className="manageNetworksHeading"
          >
            Manage Networks
          </h1>
          <CustomNetworkCard
            dropDownContent={
              <ul
                className="customNetworkDropdown"
              >
                <li>
                  <button
                    className="customNetworkDropDownButton"
                    onClick={[Function]}
                  >
                    <i
                      className="fas fa-trash"
                    />
                     Delete
                  </button>
                  <Link
                    className="customNetworkLink"
                    replace={false}
                    to="/editCustomNetwork/local"
                  >
                    <i
                      className="fas fa-edit"
                    />
                    Edit
                  </Link>
                </li>
              </ul>
            }
            key="local"
            name="local"
            url="http://127.0.0.1..."
          >
            <section
              className="customNetworkCard"
            >
              <div
                className="customNetworkColorContainer"
              >
                <div
                  className="customNetworkColor"
                />
              </div>
              <div
                className="customNetworkContainer"
              >
                <h3>
                  local
                </h3>
                <h3
                  className="customNetworkUrl"
                >
                  http://127.0.0.1...
                </h3>
              </div>
              <DropDown
                buttonContent={
                  <i
                    className="fas fa-ellipsis-v"
                  />
                }
                buttonStyles="customNetworkDropdownButton"
                classNames="customNetworkDropDown"
                dropDownContent={
                  <ul
                    className="customNetworkDropdown"
                  >
                    <li>
                      <button
                        className="customNetworkDropDownButton"
                        onClick={[Function]}
                      >
                        <i
                          className="fas fa-trash"
                        />
                         Delete
                      </button>
                      <Link
                        className="customNetworkLink"
                        replace={false}
                        to="/editCustomNetwork/local"
                      >
                        <i
                          className="fas fa-edit"
                        />
                        Edit
                      </Link>
                    </li>
                  </ul>
                }
              >
                <section
                  className="dropDownContainer customNetworkDropDown"
                >
                  <button
                    className="dropDownButton customNetworkDropdownButton"
                    onClick={[Function]}
                  >
                    <i
                      className="fas fa-ellipsis-v"
                    />
                  </button>
                  <div
                    className="dropDown undefined"
                  >
                    <ul
                      className="customNetworkDropdown"
                    >
                      <li>
                        <button
                          className="customNetworkDropDownButton"
                          onClick={[Function]}
                        >
                          <i
                            className="fas fa-trash"
                          />
                           Delete
                        </button>
                        <Link
                          className="customNetworkLink"
                          replace={false}
                          to="/editCustomNetwork/local"
                        >
                          <a
                            className="customNetworkLink"
                            href="/editCustomNetwork/local"
                            onClick={[Function]}
                          >
                            <i
                              className="fas fa-edit"
                            />
                            Edit
                          </a>
                        </Link>
                      </li>
                    </ul>
                  </div>
                </section>
              </DropDown>
            </section>
          </CustomNetworkCard>
        </section>
      </div>
    </CustomNetworkList>
  </Router>
</StaticRouter>
`;
